# Fake News Detection using machine learning in python

## Importing important libraries
#### We start with importing all the important libraries used in this project

In [2]:
import numpy as np
import pandas as pd
import itertools
from sklearn.model_selection import train_test_split
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.linear_model import PassiveAggressiveClassifier
from sklearn.metrics import accuracy_score, confusion_matrix

## Importing fake news dataset

#### Download the dataset from <a href = "https://drive.google.com/file/d/1D11a8LkjYm8s3NToy-XvnQLmHF0Q11CI/view?usp=sharing ">here</a>

In [3]:
data = pd.read_csv(r"E:\Projects\Fake news detection\news\news.csv")

In [4]:
print(data.shape)
data.head()

(6335, 4)


Unnamed: 0                                              title  \
0        8476                       You Can Smell Hillary’s Fear   
1       10294  Watch The Exact Moment Paul Ryan Committed Pol...   
2        3608        Kerry to go to Paris in gesture of sympathy   
3       10142  Bernie supporters on Twitter erupt in anger ag...   
4         875   The Battle of New York: Why This Primary Matters   

                                                text label  
0  Daniel Greenfield, a Shillman Journalism Fello...  FAKE  
1  Google Pinterest Digg Linkedin Reddit Stumbleu...  FAKE  
2  U.S. Secretary of State John F. Kerry said Mon...  REAL  
3  — Kaydee King (@KaydeeKing) November 9, 2016 T...  FAKE  
4  It's primary day in New York and front-runners...  REAL

## Looking at the data

#### Let's look at the very first news, which is fake.

In [15]:
data.iloc[0,1]

'You Can Smell Hillary’s Fear'

In [13]:
data.iloc[0,2]

'Daniel Greenfield, a Shillman Journalism Fellow at the Freedom Center, is a New York writer focusing on radical Islam. \nIn the final stretch of the election, Hillary Rodham Clinton has gone to war with the FBI. \nThe word “unprecedented” has been thrown around so often this election that it ought to be retired. But it’s still unprecedented for the nominee of a major political party to go war with the FBI. \nBut that’s exactly what Hillary and her people have done. Coma patients just waking up now and watching an hour of CNN from their hospital beds would assume that FBI Director James Comey is Hillary’s opponent in this election. \nThe FBI is under attack by everyone from Obama to CNN. Hillary’s people have circulated a letter attacking Comey. There are currently more media hit pieces lambasting him than targeting Trump. It wouldn’t be too surprising if the Clintons or their allies were to start running attack ads against the FBI. \nThe FBI’s leadership is being warned that the entir

In [5]:
labels=data.label
labels.head()

0    FAKE
1    FAKE
2    REAL
3    FAKE
4    REAL
Name: label, dtype: object

## Splitting the dataset into training and testing sets

In [6]:
x_train,x_test,y_train,y_test=train_test_split(data['text'], data['label'], test_size=0.2)

#### We now initialize a TfidfVectorizer with stop words from the English language and a maximum document frequency of 0.7 (terms with a higher document frequency will be discarded). The TfidfVectorizer turns a collection of raw documents into a matrix of TF-IDF features.

#### Now, fit and transform the vectorizer on the train set, and transform the vectorizer on the test set

In [7]:
tfidf_vectorizer=TfidfVectorizer(stop_words='english', max_df=0.7)
tfidf_train=tfidf_vectorizer.fit_transform(x_train) 
tfidf_test=tfidf_vectorizer.transform(x_test)

#### Next, we’ll initialize a PassiveAggressiveClassifier and fit this on tfidf_train and y_train. And then predict on the test set from the TfidfVectorizer and calculate the accuracy.

In [8]:
pac=PassiveAggressiveClassifier(max_iter=50)
pac.fit(tfidf_train,y_train)
y_pred=pac.predict(tfidf_test)
score=accuracy_score(y_test,y_pred)
print("Accuracy is equal to {}".format(round(score*100,3)))

Accuracy is equal to 93.291


#### We get the accuracy of 93.3% with this model. Let’s also print out confusion matrix to find the number of false & true negatives & positives.

In [9]:
confusion_matrix(y_test,y_pred, labels=['FAKE','REAL'])

array([[594,  44],
       [ 41, 588]], dtype=int64)

#### So, we get 594 true positives, 588 true negatives, 41 false positives, and 44 false negatives.

# Conclusion

#### We took a political dataset, implemented TfidfVectorizer, initialized PassiveAggressiveClassifier, and fit model. We got an accuracy of 93.3%. 